[decisions] extra libraries (#4245)
ElektraInitiative/libelektra: doc/decisions/library_split.md

## Decision

> Also allow `libelektra-extra` (and maybe others) to access private Key/KeySet.

Reviewer: ```suggestion
Also allow `libelektra-extra` (and maybe other explicitly documented libraries) to access private Key/KeySet.
```

## Implications

- we need to clearly communicate which plugins must be exactly in the version of the `libelektra-core`
- all libraries will share a versioning scheme and are only supported if used in the same version

## Related Decisions

- [Lowlevel library](lowlevel_library.md)
- [Extra library](extra_library.md)

## Notes

- `libelektra-core`:
  The core minimal API of Elektra.
  Defines what `ElektraKey` and `ElektraKeyset` are, and contains the minimal API for manipulating them.

  **Prefixes:** `elektraKey*` and `elektraKeyset*`

- `libelektra-kdb`:
  The main API for interfacing with the KDB.

  **Prefix:** `elektraKdb*`

- `libelektra-lowlevel-c`:
  Additional C APIs that are useful when working with `ElektraKey` and `ElektraKeyset`, but not are not minimal API.
  Specifically targets C and not intended for use via bindings.

>   **Prefix:** `elektraCKey*`, `elektraCKeyset*`

Reviewer: ```suggestion
  **Prefixes:** `elektraCKey*`, `elektraCKeyset*`
```

- `libelektra-highlevel-c`:
  The C high-level API for reading/writing configuration with Elektra.
  Intended for use by applications.
  Partially intended for use with code-generation.
  Specifically targets C and not intended for use via bindings.

  **Prefix:** `elektraHlc*`
  (Note: Even though there isn't any other `highlevel` implementation, `c` suffix to show this lib is only for C)

- `libelektra-opts`:
  The API for parsing command-line arguments according to Elektra's spec.

>   **Prefix:** `elektraOpts*`

Reviewer: ```suggestion

  **Prefix:** `elektraOpts*`
```

  **Prefix:** `elektraMerge*`

- `libelektra-plugin`:
  The plugin API.
  It forms the base for all plugins and contains the functions required to implement a plugin.
  It also contains the API for interacting with (other) plugins.
  This includes loading plugins and calling exported functions.

  Note: This library comes with two separate headers.
  One for implementing a plugin and another for loading/calling plugins.

  **Prefix:** `elektraPlugin*`
  (Note: previously `libelektra-plugin` merged with `libelektra-invoke`)

- `libelektra-pluginload`:
  Internal static library, linked into `libelektra-plugin` (re-exported) and `libelektra-kdb` (not exported).
  Contains the code for loading plugins.

  **Prefix:**: `elektraPlugin*` (shared with `libelektra-plugin` because exported there)

- `libelektra-mount`:
  The API for manipulating mountpoints.

  **Prefix:** `elektraMount*`

- `libelektra-type`:
  The API that defines Elektra's type system.

  **Names:** `elektra<TYPE>ToString` and `elektraKeyTo<TYPE>`
  (Note: extracted from `libelektra-ease`)

- `libelektra-extra`:
  Contains extra APIs for `ElektraKey` and `ElektraKeyset` beyond the minimal API of `libelektra-core`.

>   These APIs are things that could be considered part of the "classes" for `ElektraKey` and `ElektraKeyset`, but which we do not consider minimal.

Reviewer: ```suggestion
  These APIs are things that could be considered part of the classes for `ElektraKey` and `ElektraKeyset`, but which we do not consider minimal.
```

  The APIs should not specifically target C (see `libelektra-lowlevel-c` for that) and should be usable via bindings (if appropriate for the other language).

>   **Prefix:** `elektraExtra*`

Reviewer: Where are the array functions? Wouldn't they be called `elektraArray`?

Author: I'd just use `elektraExtraArray*` or maybe things like `elektraExtraKeyIsArrayName` and `elektraExtraKeysetGetArray`. Depends really what kind of array functions we'll have.
